<a href="https://colab.research.google.com/github/GOPIKAV77/Python/blob/main/keras_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

The given code is to generate a convolutional neural network. Keras ia an opensource deep learning framework designed to simplify the development of neural networks.

At first layers and models are imported from keras. Layers module define various neural network layers and models module create the architecture.

Sequential is used to add linear stack of layers. when  we built the model we add the layers equentially.

**model.add(layers.Conv2D(32,(3,3),activation = 'relu', input_shape = (28,28,1)))**:  In the convolutional layer we add 32 different kernel and with a size of (3,3) which means that it is a matrix of size 3x3.

Here ReLu activation is used which helps to introduce non-linearity and prevents vanishing gradients. The input will be of the shape (28,28,1) where 28 is the height of image , 28 is the width of the image and 1 is the number of chanels.

Max Pooling is downsampling operations that reduces the spatial dimensions of feature maps while retaining most important features. Here a 2x2 pooling window is used. This helps to prevent overfitting.


In [1]:
from keras import layers
from keras import models

model = models.Sequential()
model.add(layers.Conv2D(32,(3,3),activation = 'relu', input_shape = (28,28,1)))

model.add(layers.MaxPooling2D((2,2)))
model.add(layers.Conv2D(64,(3,3), activation = 'relu'))

model.add(layers.MaxPooling2D((2,2)))
model.add(layers.Conv2D(64,(3,3),activation='relu'))

/usr/local/lib/python3.11/dist-packages/keras/src/layers/convolutional/base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [2]:
model.summary()

Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ conv2d (Conv2D)                      │ (None, 26, 26, 32)          │             320 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ max_pooling2d (MaxPooling2D)         │ (None, 13, 13, 32)          │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ conv2d_1 (Conv2D)                    │ (None, 11, 11, 64)          │          18,496 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ max_pooling2d_1 (MaxPooling2D)       │ (None, 5, 5, 64)            │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ conv2d_2 (Conv2D)                    │ (None, 3, 3, 64)            │          36,928 │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 55,744 (217.75 KB)

 Trainable params: 55,744 (217.75 KB)

 Non-trainable params: 0 (0.00 B)

The formula used to get the Param values is : **(Filter Width x Filter Height x Input Channels + Bias ) x Number of filters.**



*   (3 x 3 x 1 + 1) x 32 = 320
*   Input: (28, 28, 1)
*   28-3+1 = 26




In [3]:
model.add(layers.Flatten())
model.add(layers.Dense(64, activation = 'relu'))
model.add(layers.Dense(10, activation = 'softmax'))

Flatten coverts 3D output from the last Conv2D layer into a 1D vector. The Dense layer is a fully connected layer where each neuron in the layer is connected to every neuron in the previous layer. The output size is 64 for the next layer the input size is 64 and output size is 10.

In [4]:
model.summary()

Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ conv2d (Conv2D)                      │ (None, 26, 26, 32)          │             320 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ max_pooling2d (MaxPooling2D)         │ (None, 13, 13, 32)          │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ conv2d_1 (Conv2D)                    │ (None, 11, 11, 64)          │          18,496 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ max_pooling2d_1 (MaxPooling2D)       │ (None, 5, 5, 64)            │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ conv2d_2 (Conv2D)                    │ (None, 3, 3, 64)            │          36,928 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ flatten (Flatten)                    │ (None, 576)                 │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense (Dense)                        │ (None, 64)                  │          36,928 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_1 (Dense)                      │ (None, 10)                  │             650 │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 93,322 (364.54 KB)

 Trainable params: 93,322 (364.54 KB)

 Non-trainable params: 0 (0.00 B)

In [5]:
from keras.datasets import mnist
from keras.utils import to_categorical

(train_images, train_labels),(test_images, test_labels) = mnist.load_data()

train_images = train_images.reshape((60000, 28,28, 1))
train_images = train_images.astype('float32')/255

test_images = test_images.reshape((10000, 28, 28,1))
test_images = test_images.astype('float32')/255

train_labels = to_categorical(train_labels)
test_labels = to_categorical(test_labels)

model.compile(optimizer = 'rmsprop',loss = 'categorical_crossentropy',metrics = ['accuracy'])
model.fit(train_images, train_labels, epochs = 5, batch_size = 64)

11490434/11490434 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step
Epoch 1/5
938/938 ━━━━━━━━━━━━━━━━━━━━ 80s 80ms/step - accuracy: 0.8639 - loss: 0.4314
Epoch 2/5
938/938 ━━━━━━━━━━━━━━━━━━━━ 55s 59ms/step - accuracy: 0.9840 - loss: 0.0520
Epoch 3/5
938/938 ━━━━━━━━━━━━━━━━━━━━ 73s 49ms/step - accuracy: 0.9892 - loss: 0.0316
Epoch 4/5
938/938 ━━━━━━━━━━━━━━━━━━━━ 84s 52ms/step - accuracy: 0.9931 - loss: 0.0226
Epoch 5/5
938/938 ━━━━━━━━━━━━━━━━━━━━ 86s 56ms/step - accuracy: 0.9944 - loss: 0.0179


The to_categorical converts class labels into one-hot encoded vectors, which are required for multi-class classification problems in meural network.

for exmaple if the number is 7 the [0,0,0,0,0,0,0,1,0,0]

Pixel values range from 0 to 255. Dividing by 255 helps in normalising the value and it ranges from 0 to 1.

For compiling the model, rmsprop (Root mean square propagation) is used for adaptive learning. categorical crossentropy is used for multi class classification. accuracy is used to measure how many predictions are correct.

In [6]:

test_loss, test_acc = model.evaluate(test_images,test_labels)
test_acc

313/313 ━━━━━━━━━━━━━━━━━━━━ 3s 9ms/step - accuracy: 0.9898 - loss: 0.0385


0.9927999973297119